src/btree/bt_read.c

```diff
@@ -247,11 +247,23 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
                 return (WT_NOTFOUND);
             goto read;
         case WT_REF_DISK:
-            /* Limit reads to cache-only, or internal pages only. */
-            if (LF_ISSET(WT_READ_CACHE) ||
-              (LF_ISSET(WT_READ_CACHE_LEAF) && __wt_ref_is_leaf(session, ref)))
+            /* Limit reads to cache-only */
+            if (LF_ISSET(WT_READ_CACHE))
                 return (WT_NOTFOUND);
 
+            /* Limit reads to internal pages only. */
+            if (LF_ISSET(WT_READ_CACHE_LEAF)) {
+                /* Lock the ref to avoid it getting changed, while we check the page type. */
+                if (!WT_REF_CAS_STATE(session, ref, WT_REF_DISK, WT_REF_LOCKED))
+                    return (WT_NOTFOUND);
+
```

Comment:
I suggest removing this blank line because it then keeps all the lock/op/unlock lines together.

Reply:
The fact we were unable to set the state to `WT_REF_LOCKED` doesn't tell us anything about whether or not the page is a leaf page.

Reply:
I agree that failure in state change to `WT_REF_LOCKED` led to page read failure. I added a comment explaining the scenario where the `WT_READ_CACHE_LEAF` is used and what happens in case of a failure.

Reply:
If we fail to CAS into a locked state, why not restart the loop? The state of the page has changed, why not retry the read? I'm suggesting something like:

```
*** bt_read.c   Thu Jan 23 00:08:42 2020
--- -   Thu Jan 23 00:11:58 2020
*************** 
*** 267,273 ****
                   * a disk page that can led to a failure.
                   */
                  if (!WT_REF_CAS_STATE(session, ref, WT_REF_DISK, WT_REF_LOCKED))
!                     return (WT_NOTFOUND);

                  is_leaf_page = __wt_ref_is_leaf(session, ref);
                  WT_REF_SET_STATE(ref, WT_REF_DISK);
--- 267,273 ----  
                   * a disk page that can led to a failure.
                   */
                  if (!WT_REF_CAS_STATE(session, ref, WT_REF_DISK, WT_REF_LOCKED))
!                     continue;

                  is_leaf_page = __wt_ref_is_leaf(session, ref);
                  WT_REF_SET_STATE(ref, WT_REF_DISK);
```
That way we don't have to care too much about how the function was called.

Reply:
A retry is a nice approach, Fixed.